### DISCUSSION TITLE
Programmatic run of uvicorn.run forks many processes causes k8s pod restarts

### DISCUSSION THREAD

**Discussion by timothy-mugayi** [fastapi/fastapi#6755] [Questions]
Hi guys I seem to be having some issues probably my lack of understanding i do apologize if issue may seem trivial am running below 

command to run fast API programmatically 

`uvicorn.run("audio_service.worker:app", host="localhost", port=8000, reload=False, workers=1, debug=True, log_level="trace", access_log=False)`

I have also set the following 

```
ENV WORKERS_PER_CORE=2
ENV WEB_CONCURRENCY=1
```
within my docker container.  Running tensorflow to serve a model via an REST API endpoint nothing special here just a binary true false after inference. 

i installed htop and what i do see is 

/opt/conda/bin/python -B -c from multiprocessing spawn import spawn_main; spawn_main(tracker_fd=7, pipe_handle=9) --multiprocessing-fork root

followed by a list of processes that keep getting forked maxing out the total CPU cores. Is they any way to limit this so that we can use 1 or 2 workers? The above configurations don't seem to be taking any effect. 

This leads to 

---CUT---
    State:          Running
      Started:      Tue, 30 Mar 2021 13:53:24 -0700
    Last State:     Terminated
      Reason:       Error
      Exit Code:    137
      Started:      Tue, 30 Mar 2021 13:32:34 -0700
      Finished:     Tue, 30 Mar 2021 13:53:23 -0700

**Comment by MartinEliasQ** [comment]
Hello @timothy-mugayi 
Have you tried to increase the POD resource limits?
In any case, to make the deployments more atomic I use a concurrency of 1 for gunicorn workers on uvicorn. When you upload your model it will for sure go to RAM and it is likely that it will not finish booting due to lack of resources.

**Comment by timothy-mugayi** [comment]
@MartinEliasQ  i have tried below which seems to have reduced the total number of forked processes 

`uvicorn.run("audio_service.worker:app", host="0.0.0.0", port=8002, reload=False, workers=1, debug=False, log_level="trace", access_log=False)`

ENV WORKERS_PER_CORE=1
ENV WEB_CONCURRENCY=1

htop we get 

   ```
 1 root       20   0  438M  118M 43944 S  0.0  1.2  0:16.77 python -m audio_service.worker --broker ym_api
  315 root       20   0  5052  3344  2876 R  0.0  0.0  0:04.34 htop
   10 root       20   0  438M  118M 43944 S  0.0  1.2  0:00.00 python -m audio_service.worker --broker ym_api
   11 root       20   0  438M  118M 43944 S  0.0  1.2  0:00.00 python -m audio_service.worker --broker ym_api
   12 root       20   0  438M  118M 43944 S  0.0  1.2  0:00.00 python -m audio_service.worker --broker ym_api
   13 root       20   0  438M  118M 43944 S  0.0  1.2  0:00.00 python -m audio_service.worker --broker ym_api
   14 root       20   0  5752  3684  3112 S  0.0  0.0  0:00.06 bash
```
I noticed setting debug=False turns of the auto-reloading while reload=False didn't do much

before doing this it kept on forking as many processes as available on the machine.

**Comment by raphaelauv** [comment]
Reduce your program to a minimal running fastapi endpoint without any call to other code and loading any other lib, you will see that it's very probably your code that is responsible. If not then share with us a minimal reproducible example.

**Comment by raphaelauv** [comment]
@timothy-mugayi could you please close the issue , thanks

**Comment by aucampia** [comment]
@timothy-mugayi issue still open, probably should be closed.
